python/graphstorm/model/loss_func.py

```diff
@@ -172,3 +207,68 @@ def out_dims(self):
         int : the number of output dimensions.
         """
         return None
+
+class LinkPredictContrastiveLossFunc(GSLayer):
+    """ Contrastive Loss function for link prediction.
+
+        Parameters
+        ----------
+        temp: float
+            Temperature value
+    """
+    def __init__(self, temp=1.0):
+        super(LinkPredictContrastiveLossFunc, self).__init__()
+        self._temp = temp
+
+    def forward(self, pos_score, neg_score):
+        """ The forward function.
+
+            Parameters
+            ----------
+            pos_score: dict of Tensor
+                The scores for positive edges of each edge type.
+            neg_score: dict of Tensor
+                The scores for negative edges of each edge type.
+        """
+        pscore = []
+        nscore = []
+        for key, p_s in pos_score.items():
+            assert key in neg_score, \
+                f"Negative scores of {key} must exists"
+            n_s = neg_score[key]
+
+            # Both p_s and n_s are soreted according to source nid
+            # (which are same in pos_graph and neg_graph)
+            pscore.append(p_s)
+            nscore.append(n_s.reshape(p_s.shape[0], -1))
+        pscore = th.cat(pscore, dim=0)
+        nscore = th.cat(nscore, dim=0)
+
+        pscore = th.div(pscore, self._temp)
+        nscore = th.div(nscore, self._temp)
+        score = th.cat([pscore.unsqueeze(1), nscore], dim=1)
+
+        exp_logits = th.exp(score)
+        log_prob = pscore - th.log(exp_logits.sum(1))
```

Comment:
can you write the formula of the contrastive loss? it's not very clear to me.

Reply:
Added in the class description